### PR TITLE
[Apple] Fix flaky unit test

### PR DESCRIPTION
**Pull request by waxion03**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [X] Are tests passing locally?
* [ ] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description
Reproduce the bug, try to solve it，remain to be observed.

## Related PRs or issues

List related PRs and other issues.

### CODE REVIEW

File: AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m [hunk 1]
```diff
@@ -226,7 +226,7 @@ - (void)sendCallAsync:(MSIngestionCall *)call {
     NSURLSessionDataTask *task = [self.session
         dataTaskWithRequest:request
           completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
-            @synchronized(self) {
+            @synchronized(@"sendCallAsync") {
```
Comment (Jamminroot) [inline AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: Can you explain using NSString as a monitor? It doesn't look like correct code at all.

Comment (waxion03) [reply AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: I just tried，still analyzing the problem really，from the log point of view is the crash in this method.

Comment (waxion03) [reply AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: Maybe a thread problem.

File: AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m [hunk 2]
```diff
@@ -222,11 +222,13 @@ - (void)sendCallAsync:(MSIngestionCall *)call {
       return;
     }
 
+    // Create a task for the request.
+    __block NSLock *taskLock = [[NSLock alloc] init];
```
Comment (Jamminroot) [inline AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: This code looks incorrect to me. I think what you do is basically create lock each time (each time code is here), so for each time it is referred on line #231, reference goes to separate monitor, which is meaningless. Correct me if im wrong.

File: AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m [hunk 3]
```diff
@@ -261,7 +263,8 @@ - (void)sendCallAsync:(MSIngestionCall *)call {
                 call.submitted = NO;
                 [call ingestion:self callCompletedWithResponse:httpResponse data:data error:error];
               }
-            }
+              [taskLock unlock];
+              taskLock = nil;
```
Comment (Jamminroot) [inline AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: Well, since its `malloc`, I think it should rather be `free`, not `= nil`

**Comment by jaelim-ms** [comment]
@waxion03 Do you think this isn't a unit test problem but a bug or defect in the SDK? If that's true than we can fix it in the SDK but the problem might be in the unit test.

File: AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m (hunk 2 above)
Comment (waxion03) [reply AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: The phenomenon I observed was that every time a test project was run, a crash might occur.and the crash happens in this function, but run a test alone will not crash, so I think this is a problem with SDK concurrency, data confusion and data security issues when multiple threads access the same resource.Correct me if im wrong.

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 4]
```diff
@@ -641,6 +644,8 @@ - (void)testShowConfirmationAlertForMandatoryUpdateWhileNoNetwork {
   NSString *message = [NSString stringWithFormat:MSDistributeLocalizedString(@"MSDistributeAppUpdateAvailableMandatoryUpdateMessage"),
                                                  appName, details.shortVersion, details.version];
 #pragma clang diagnostic pop
+    id appCenterMock = OCMClassMock([MSAppCenter class]);
```
Comment (Jamminroot) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: Looks unformatted

File: AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m [hunk 5]
```diff
@@ -221,12 +221,12 @@ - (void)sendCallAsync:(MSIngestionCall *)call {
...
       return;
     }
-
+      
     // Create a task for the request.
     NSURLSessionDataTask *task = [self.session
         dataTaskWithRequest:request
           completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
-            @synchronized(self) {
+              @synchronized(self) {
```
Comment (MatkovIvan) [inline AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: Please format

File: AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m [hunk 6]
```diff
@@ -221,7 +221,7 @@ - (void)sendCallAsync:(MSIngestionCall *)call {
     if (!request) {
       return;
     }
-
+
```
Comment (dhei) [inline AppCenter/AppCenter/Internals/Ingestion/MSHttpIngestion.m]: Please run formatter.

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 7]
```diff
@@ -622,6 +622,9 @@ - (void)testShowConfirmationAlertForMandatoryUpdateWhileNoNetwork {
 
   // Mock Bundle.
   NSString *appName = @"Test App";
+  NSString *kMSDisServiceName = @"AppCenterDistribute";
```
Comment (dhei) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: I don't think these 3 variables are used. Remove if unused.

Comment (waxion03) [reply AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: OK!

**Comment by waxion03** [comment]
> Since it's about a flaky test, it would be nice to see that it passes N times in a row on CI. Do you have a link or screenshot for these tests?
Sorry, I'm not sure how to test N times in one row. Does the build pipeline on CI server have this function? Can you tell me how to do it? Thanks~

**Comment by waxion03** [comment]
> Since it's about a flaky test, it would be nice to see that it passes N times in a row on CI. Do you have a link or screenshot for these tests?
> Sorry, I'm not sure how to test N times in one row. Does the build pipeline on CI server have this function? Can you tell me how to do it? Thanks~

**Comment by waxion03** [comment]
阿斯顿

**Comment by waxion03** [comment]
> Since it's about a flaky test, it would be nice to see that it passes N times in a row on CI. Do you have a link or screenshot for these tests?

The problem I found was that there was no mock MSAppCenter, and a lot of tests were done after the modification, and the problem did not recur. Unfortunately, there was no screenshot of this process, but it was really measured for a long time.

**Pull request merged**